src/go/k8s/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -886,3 +943,36 @@ func disableConsoleReconciliation(console *vectorzied_v1alpha1.Console) {
 	}
 	console.Annotations[managedAnnotationKey] = NotManaged
 }
+
+func checkDeploymentsStatus(deployments []*appsv1.Deployment) (string, bool) {
+	return checkReplicasForList(func(o *appsv1.Deployment) (int32, int32, int32, int32) {
+		return o.Status.UpdatedReplicas, o.Status.AvailableReplicas, o.Status.ReadyReplicas, ptr.Deref(o.Spec.Replicas, 0)
+	}, deployments, "Deployment")
+}
+
+func checkStatefulSetStatus(ss []*appsv1.StatefulSet) (string, bool) {
+	return checkReplicasForList(func(o *appsv1.StatefulSet) (int32, int32, int32, int32) {
+		return o.Status.UpdatedReplicas, o.Status.AvailableReplicas, o.Status.ReadyReplicas, ptr.Deref(o.Spec.Replicas, 0)
+	}, ss, "StatefulSet")
+}
+
+type replicasExtractor[T client.Object] func(o T) (updated, available, ready, total int32)
+
+func checkReplicasForList[T client.Object](fn replicasExtractor[T], list []T, resource string) (string, bool) {
+	var notReady sort.StringSlice
+	for _, item := range list {
+		updated, available, ready, total := fn(item)
+
+		if updated != total || available != total || ready != total {
+			name := client.ObjectKeyFromObject(item).String()
+			item := fmt.Sprintf("%q (updated/available/ready/total: %d/%d/%d/%d)", name, updated, available, ready, total)
+			notReady = append(notReady, item)
+		}
+	}
+	if len(notReady) > 0 {
+		notReady.Sort()
+
+		return fmt.Sprintf("Not all replicas updated, available, and ready for %s(s) [%s].", resource, strings.Join(notReady, "; ")), false
```

Comment:
Not blocking comment:

```
and ready for %s(s) [%s]
```
How I should interpret `(s)`?

In real word `and ready for statefulset(s) [someName (updated/available/ready/total: 1/1/0/3)]` that s would mean optional plural for particular resource?

Reply:
Yep, it's a bit more of a future-proofing in case we move from StatefulSets two Deployments, or if for some reason we have multiple of something we have to check, I can lightly change the wording to make it work with an ambiguous number of resources. WDYT about something like this?

```
Not all StatefulSet replicas updated, available, and ready for [someName ..., someOtherName]
```

It'd work whether you have a single or multiple resources in the array.

Reply:
Sure, works for me. It's not blocking comment :)